ENH: Vectorize packbits with SSE2

juliantaylor opened the pull request:
SSE2 has a special instruction to pack bytes into bits,
available as the intrinsic _mm_movemask_epi8. It is significantly faster
than the per byte loop currently being used.
Unfortunately packbits is bitwise "big endian", the most significant bit
is the first in the input byte while _mm_movemask_epi is little endian
so we need to byteswap the input first. But it is still about 8-10 times
faster than the scalar code.

It's probably somewhat complicated code for a function which is not very important but the speedup is pretty decent and I think its reasonably well tested now (though I couldn't find a way to test inner_stride == 1 and outerstride != 1)
unpackbits could probably also be speed up with e.g. a lookup table.

juliantaylor committed:
TST: add extended packbits tests

Larger data to account for future vectorization.
Also add benchmarks for packbits and unpackbits

juliantaylor committed:
ENH: vectorize packbits with SSE2

SSE2 has a special instruction to pack bytes into bits,
available as the intrinsic _mm_movemask_epi8. It is significantly faster
than the per byte loop currently being used.
Unfortunately packbits is bitwise "big endian", the most significant bit
is the first in the input byte while _mm_movemask_epi is little endian
so we need to byteswap the input first. But it is still about 8-10 times
faster than the scalar code.

charris commented:
How does this play with #8327.

juliantaylor commented:
only a possible merge conflict in the tests, the other should be merged first, I'll sort it out then.

charris commented:
@juliantaylor Needs rebase.

homu commented:
:umbrella: The latest upstream changes (presumably #8327) made this pull request unmergeable. Please resolve the merge conflicts.

juliantaylor committed:
ENH: use a lookup table for unpack bits

Uses 2KiB of space, but makes unpack bits 8 times faster and comparable
with the vectorized packbits.

juliantaylor commented:
updated, also couldn't resist adding a lookup table for unpackbits to make its performance similar to packbits.
The table is 2KiB in size though :/

charris commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1497,13 +1501,40 @@ pack_inner(const char *inptr,
         remain = 8;
     }
-    for (index = 0; index < n_out; index++) {
+    for (; index < n_out; index++) {
```
So this picks up the index from the preceding block, otherwise has the value zero?

charris commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1704,15 +1996,24 @@ unpack_bits(PyObject *input, int axis)
-                *outptr = ((mask & (*inptr)) != 0);
-                outptr += out_stride;
-                mask >>= 1;
+                for (i = 0; i < 8; i++) {
```
I wonder if it might be faster to use a local buffer and the lookup table, then copy the result to the destination?

charris replied:
Or just copy from the table directly with the stride for the output pointer.

juliantaylor replied:
I don't think this would be faster, you need a 8 element loop anyway and the original one is pretty fast already.

charris commented on numpy/lib/tests/test_packbits.py:
```diff
@@ -51,6 +53,166 @@ def test_packbits_empty_with_axis():
+                  1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0,
+                  1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0])
+    a = a.repeat(3)
+    for dtype in [np.bool, np.uint8, np.int32, np.int64]:
```
Any particular reason for just these dtypes?

juliantaylor replied:
there isn't really much point in testing more than bool and one integer as there are only two paths, but then the test is probably fast and could get more path in future.

charris commented on numpy/lib/tests/test_packbits.py:
```diff
@@ -86,3 +248,18 @@ def test_unpackbits_empty_with_axis():
+
+
+def test_unpackbits_large():
+    # test all possible numbers via comparison to already tested packbits
```
OK.

charris commented:
Might add a mention in the release notes under Improvements.

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1497,13 +1501,40 @@ pack_inner(const char *inptr,
+    if (in_stride == 1 && element_size == 1 && n_out > 2) {
+        __m128i zero = _mm_setzero_si128();
+        /* don't handle non-full 8-byte remainder */
+        npy_uintp vn_out = n_out - (remain ? 1 : 0);
```
yes its for the modulus, probably nicer to just leave it signed and use ´&`

juliantaylor commented on numpy/core/src/multiarray/compiled_base.c:
```diff
@@ -1625,6 +1656,267 @@ pack_bits(PyObject *input, int axis)
 }
 
+/* lookup table for unpackbits */
+static char unpack_lookup[256][8] = {
```
we could also generate it on first use under the GIL

juliantaylor committed:
DOC: add packbits improvements to release notes

charris commented on doc/release/1.13.0-notes.rst:
```diff
@@ -87,6 +87,11 @@ it should be regarded as a band-aid until Mingwpy is fully functional.
 .. _issues: https://mingwpy.github.io/issues.html
 
+Performance improvements for ``packbits`` and ``unpackbits``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Just a note for the future, I've changed the `~` to `-` for added markdown compatibility for the github releases. I'll fix these up in the release notes later.

charris commented:
Thanks Julian.

pv commented:
For me this seems to introduce a crash: gh-8637
